cs/src/Connections/RelayTunnelConnector.cs

```diff
@@ -48,169 +48,220 @@ public RelayTunnelConnector(IRelayClient relayClient)
-                    ex is ArgumentException)
+                catch (SshConnectionException scex)
                 {
-                    throw;
+                    // All other SSH connection exceptions are not recoverable.
+                    disconnectReason = scex.DisconnectReason != SshDisconnectReason.None ? scex.DisconnectReason : SshDisconnectReason.ByApplication;
+                    throw new TunnelConnectionException($"Failed to start tunnel SSH session: {scex.Message}", scex);
                 }
+                catch (WebSocketException wse)
+                {
+                    if (wse.WebSocketErrorCode == WebSocketError.UnsupportedProtocol)
+                    {
+                        throw new InvalidOperationException("Unsupported web socket sub-protocol.", wse);
+                    }
 
-                errorDescription = ex.Message;
-                exception = ex;
-            }
-            finally
-            {
-                if (disconnectReason != SshDisconnectReason.None)
+                    if (wse.WebSocketErrorCode == WebSocketError.NotAWebSocket)
+                    {
+                        var statusCode = TunnelConnectionException.GetHttpStatusCode(wse);
+                        switch (statusCode)
+                        {
+                            case HttpStatusCode.Unauthorized:
+                                exception = new UnauthorizedAccessException(
+                                    $"Unauthorized (401). Provide a fresh tunnel access token with '{this.relayClient.TunnelAccessScope}' scope.",
+                                    wse);
+
+                                ThrowIfRetryNotAllowed(exception, SshDisconnectReason.AuthCancelledByUser, delayNeeded: false);
+
+                                // Unauthorized error may happen when the tunnel access token is no longer valid, e.g. expired.
+                                // Try refreshing it.
+                                await RefreshTunnelAccessTokenAsync(wse);
+                                break;
+
+                            case HttpStatusCode.Forbidden:
+                                throw new UnauthorizedAccessException(
+                                    $"Forbidden (403). Provide a fresh tunnel access token with '{this.relayClient.TunnelAccessScope}' scope.",
+                                    wse);
+
+                            case HttpStatusCode.NotFound:
+                                throw new TunnelConnectionException($"The tunnel or port is not found (404).", wse);
+
+                            case HttpStatusCode.TooManyRequests:
+                            case HttpStatusCode.ServiceUnavailable:
+                            case HttpStatusCode.BadGateway:
+                                // Normally nginx choses another healthy pod when it cannot establish connection to a pod.
+                                // However, if there are no other pods, it may returns 502 (Bad Gateway) to the client.
```

Comment:
Neelima is going to change pod behavior to about HTTP connection when nginx need to choose another pod. If nginx cannot choose any, it would return 502. The comment on line 148 explains that.